### PR TITLE
Replace makd with dub and upgrade deps

### PR DESCRIPTION
**Pull request by mihails-strasuns**
Major build system rework:

- each generated binary defined as separate DUB sub-package
- each top-level utlity library package is also defined as DUB sub-package
- tests redefined using dub configurations
- dependency versions upgraded to latest available (i.e. vibe.d 0.8)
- makd only used for packaging
- Travis script updated to match our current standards

This is one huge commit but all these changes are so subtly related that I really don't want to spend time trying to untangle this :(

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -11,56 +11,46 @@ services:
...
 
-# Do a shallow submodule fetch
-before_install: git submodule update --init
-
+# Global environment variables
 env:
     global:
         - DIST=xenial
-        - ARTIFACTS_DIR=build/travis-artifacts
+        # Default beaver image names. May be overriden in specific stages
```
Comment (leandro-lucarella-sociomantic) [inline .travis.yml:18]: Are the changes to travis really that hard to split to a separate commit?

File: Build.mak [hunk 2]
```diff
@@ -1,58 +1,33 @@
...
+	dub build :overview
 
-# Compile all vibe.d as a single static library because rdmd can't track
-# all imported dependencies correctly in this case. Static library approach
-# instead of compiling all sources in one go was chosen to reduce recompilation
-# time for actual apps in this repo.
-$O/libvibed.a: $(shell find $C/submodules/vibed/source -type f -name '*.d')
-	$(call exec, dmd $(DFLAGS) -lib -of$@ $?)
+$B/neptune-autopr:
+	dub build :autopr
```
Comment (leandro-lucarella-sociomantic) [inline Build.mak]: You can probably write an implicit rule, something like:

```make
tools := overview release autopr

all += $(patsubst %,$B/neptune-%,$(tools))

# Recompile on flag changes
$B/%: $G/build-d-flags

# Build with dub
$B/neptune-%: $C/src/%/main.d
    dub build :$*
```

File: Build.mak [hunk 3]
```diff
@@ -1,58 +1,33 @@
...
+all += $B/neptune-release $B/neptune-autopr $B/neptune-overview
 
-################################################################################
+# Wrapper for dub test to be used from CI
 
-$O/allunittests $O/fastunittest: $O/libvibed.a $O/libdyaml.a $O/libd-colorize.a
+dub-unittest:
+	dub test :overview
+	dub test :release
+	dub test :autopr
```
Comment (leandro-lucarella-sociomantic) [inline Build.mak]: Similarly:

```make
$O/unittest-%.stamp: $C/src/%/main.d
    dub test :$*
.PHONY: dub-unittest
dub-unittest: $(patsubst %,$O/unittest-%,$(tools))
```

But I think you could even redefine the `allunittest` target or something.

File: Build.mak [hunk 4]
```diff
@@ -1,58 +1,33 @@
...
+ITFLAGS := --tmp=$O --bin=$B
 
-$O/test-first_release: $B/neptune-release $O/libvibed.a
-$O/test-prerelease: $B/neptune-release $O/libvibed.a
-$O/test-preview_release: $B/neptune-release $O/libvibed.a
+dub-inttest: all
+	dub run :tests -c first_release   -- $(ITFLAGS)
+	dub run :tests -c prerelease      -- $(ITFLAGS)
+	dub run :tests -c patch_release   -- $(ITFLAGS)
+	dub run :tests -c preview_release -- $(ITFLAGS)
```
Comment (leandro-lucarella-sociomantic) [inline Build.mak]: And same again:

```make
$O/test-%.stamp: $C/$(INTEGRATIONTEST)/%/main.d
    dub run :tests -c $*   -- $(ITFLAGS)

.PHONY: dub-inttest
dub-inttest: $(patsubst $(INTEGRATIONTEST)/%/main.d,$O/test-%.stamp,$(wildcard $(INTEGRATIONTEST)/*/main.d))
```

(all this untested, of course, but just so you get the idea :) )

File: Build.mak [hunk 5]
```diff
@@ -1,58 +1,33 @@
...
-
-$B/neptune-release: override LDFLAGS += -L$O -lvibed -levent -lssl -lcrypto -ld-colorize
-$B/neptune-release: $C/src/release/main.d $O/libvibed.a $O/libd-colorize.a
-
-release: $B/neptune-release
-overview: $B/neptune-overview
-autopr: $B/neptune-autopr
-
-all += release overview autopr
+dub-test: dub-unittest dub-inttest
```
Comment (leandro-lucarella-sociomantic) [inline Build.mak:30]: Why the `dub-*`instead of the regular makd targets?

File: Makefile [hunk 6]
```diff
@@ -5,3 +5,6 @@ include submodules/makd/Makd.mak
 # Default goal for building this directory
 .DEFAULT_GOAL := all
 
+# DUB doesn't work well with flavors but there isn't any need in those
+# for neptune apps anyway:
+VALID_FLAVORS := dub
```
Comment (leandro-lucarella-sociomantic) [inline Makefile]: ah... So you want to make both available concurrently? :thinking:

File: src/internal/git/helper.d [hunk 7]
```diff
@@ -10,9 +10,9 @@
 
 *******************************************************************************/
 
-module lib.git.helper;
+module internal.git.helper;
```
Comment (leandro-lucarella-sociomantic) [inline src/internal/git/helper.d:13]: Why the name change from `lib` to `internal`?

File: src/octod/api/authorizations.d [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    Provides wrappers on top of some methods documented in
+    https://developer.github.com/v3/oauth_authorizations/
+
+    Copyright: Copyright (c) 2016 sociomantic labs GmbH. All rights reserved
+
+    License: Boost Software License Version 1.0 (see LICENSE for details)
+
+*******************************************************************************/
+
+module octod.api.authorizations;
```
Comment (leandro-lucarella-sociomantic) [inline src/octod/api/authorizations.d:12]: What? You are also moving octod here in this same commit? :scream:

**Review by leandro-lucarella-sociomantic — CHANGES_REQUESTED** [review]
Changes look goodish to me in general, I think if we move forward in this direction eventually we could even create a `MakDub.mak` file in `makd` to use the same conventions as in `Makd.mak` but with these adaptions to dub.

What I think is a bit unacceptable is the mess of one commit to rule them all. I can understand some parts might be too interrelated and hard to split, but I think at least the octod part (which is even undocumented in the PR description) and the travis changes should be split. I think even if we ended up with some commit that doesn't build it would be better than this huge commit.

File: Build.mak (hunk 3 above)
Comment (mihails-strasuns) [reply Build.mak]: I tried that, but it looks like makd tries to do some automatic machinery despite that which fails (allunittests.d generation). Maybe not looked hard enough.

File: Build.mak (hunk 5 above)
Comment (mihails-strasuns) [reply Build.mak:30]: They are defined in makd with `:=` :(

File: Makefile (hunk 6 above)
Comment (mihails-strasuns) [reply Makefile]: Unless I am missing something obvious, dub doesn't support different output directories per build config (and it defines about 10 standard build presets, not just debug and release). As we don't care that much for neptune which one is used I decided to just use the latest (they overwrite each other).

File: src/internal/git/helper.d (hunk 7 above)
Comment (mihails-strasuns) [reply src/internal/git/helper.d:13]: To make it more obvious that this is not user-facing library package and is not exposed via `subPackage` intentionally. I plan to give a try to refactoring/restructuring this all a bit after initial PR in an attempt to make every top-level package "public".

File: src/octod/api/authorizations.d (hunk 8 above)
Comment (mihails-strasuns) [reply src/octod/api/authorizations.d:12]: Yup, this is not strictly a prerequisite, but makes things much more simple dub.sdl definition-wise.

I will split it into separate one.

File: .travis.yml (hunk 1 above)
Comment (mihails-strasuns) [reply .travis.yml:18]: If you don't mind that it ends up with CI-failing intermediate commits - no :)

File: Build.mak (hunk 2 above)
Comment (mihails-strasuns) [reply Build.mak]: Will try. Btw I don't use `.d` files as dependency intentionally here - dub does dependency tracking on its own (so running `dub build` again will be no-op) and it is much more convoluted than what I can possibly implement in a makefile.

**Comment by mihails-strasuns** [comment]
@leandro-lucarella-sociomantic 

>  I think if we move forward in this direction eventually we could even create a MakDub.mak file in makd to use the same conventions as in Makd.mak but with these adaptions to dub.

I think that at the very least there is a solid case to split packaging related bits of makd into own `MakPkg.mak` so that it can be used in any project (even non-D) without any other MakD conventions coming into effect.

File: Build.mak (hunk 3 above)
Comment (leandro-lucarella-sociomantic) [reply Build.mak]: I don't know exactly what you tried, but in theory this should work because there is no `$O/unittest-%.stamp` defined in Makd. Maybe it could have failed for `$O/test-%.stamp` because that's already defined so they might compete. If you change that to `$O/inttest-%.stamp` for now I think it should work too.

File: src/internal/git/helper.d (hunk 7 above)
Comment (leandro-lucarella-sociomantic) [reply src/internal/git/helper.d:13]: This also really looks like it should go in a separate commit, it shouldn't be that hard.

**Comment by mihails-strasuns** [comment]
Updated:

- split into several commits (low hanging fruits)
- more generalized makefile definition
- `dub -q` calls (quite, warnings/errors only) to make travis log more readable

**Comment by mihails-strasuns-sociomantic** [comment]
Ping @leandro-lucarella-sociomantic

**Review by leandro-lucarella-sociomantic — APPROVED** [review]
Thanks, looks much better. Now I don't get why the rename `integrationtest` -> `tests`, but other than that LGTM.

**Comment by mathias-baumann-sociomantic** [comment]
I checked out the branch and tried `make test` with the following result:


```
 marenz@labs-081  ~/sociomantic/neptune   pr-223  make test                                                                     :-) ~/sociomantic/neptune 
   gen build/dub/tmp/allunittests.d
   rdmd build/dub/tmp/allunittests
./src/octod/core.d(68): Error: no identifier for declarator Connection
./src/octod/core.d(68): Error: alias cannot have initializer
./src/octod/core.d(103): Error: found 'import' instead of statement
./src/octod/core.d(124): Error: found 'import' instead of statement
./src/octod/core.d(131): Error: found '!' when expecting ';' following statement
./src/octod/core.d(142): Error: found 'import' instead of statement
./src/octod/core.d(143): Error: found '!' when expecting ';' following statement
./src/octod/core.d(143): Error: found '..' when expecting ']'
./src/octod/core.d(143): Error: found '$' when expecting ')'
./src/octod/core.d(143): Error: semicolon expected to close declaration, not ']'
./src/octod/core.d(143): Error: found ']' instead of statement
./src/octod/core.d(201): Error: basic type expected, not scope
./src/octod/core.d(201): Error: found 'scope' when expecting ')'
./src/octod/core.d(201): Error: semicolon expected following function declaration
./src/octod/core.d(201): Error: found 'request' when expecting ',' following argument
./src/octod/core.d(201): Error: expression expected, not ')'
./src/octod/core.d(201): Error: found '{' when expecting ',' following argument
./src/octod/core.d(202): Error: found ';' when expecting ',' following argument
./src/octod/core.d(203): Error: found ';' when expecting ',' following argument
./src/octod/core.d(204): Error: found ';' when expecting ',' following argument
./src/octod/core.d(205): Error: expression expected, not '}'
Failed: ["dmd1", "-di", "-g", "-I/home/marenz/sociomantic/neptune/build/dub/include", "-I./src", "-I./submodules/makd/src", "-I./submodules/beaver/src", "-unittest", "-debug=UnitTest", "-version=UnitTest", "-v", "-o-", "/home/marenz/sociomantic/neptune/build/dub/tmp/allunittests.d", "-I/home/marenz/sociomantic/neptune/build/dub/tmp"]
submodules/makd/Makd.mak:595: recipe for target '/home/marenz/sociomantic/neptune/build/dub/tmp/allunittests' failed
make: *** [/home/marenz/sociomantic/neptune/build/dub/tmp/allunittests] Error 1
```

**Comment by mathias-baumann-sociomantic** [comment]
Using `DMD64 D Compiler v2.081.1`

**Comment by mihails-strasuns** [comment]
`make test` won't work because of current limitations in `makd` (I can't disable all automatic targets it implies). Because of that new `make dub-test` was added for CI purpose.

**Comment by mihails-strasuns** [comment]
Thanks!
